### DISCUSSION TITLE
Couldn't retrieve remote JWK set: org.springframework.web.client.HttpClientErrorException$Forbidden: 403

### DISCUSSION THREAD

**Discussion by pratikpanchal4472** [keycloak/keycloak#21191] [Q&A]
I am getting this error while following the demo from this medium article.
https://medium.com/geekculture/using-keycloak-with-spring-boot-3-0-376fa9f60e0b

GET http://localhost:8080/realms/supr/.well-known/openid-configuration
Response:
`{"issuer":"http://localhost:8080/realms/supr","authorization_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/auth","token_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/token","introspection_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/token/introspect","userinfo_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/userinfo","end_session_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/logout","frontchannel_logout_session_supported":true,"frontchannel_logout_supported":true,"jwks_uri":"http://localhost:8080/realms/supr/protocol/openid-connect/certs","check_session_iframe":"http://localhost:8080/realms/supr/protocol/openid-connect/login-status-iframe.html","grant_types_supported":["authorization_code","implicit","refresh_token","password","client_credentials","urn:ietf:params:oauth:grant-type:device_code","urn:openid:params:grant-type:ciba"],"acr_values_supported":["0","1"],"response_types_supported":["code","none","id_token","token","id_token token","code id_token","code token","code id_token token"],"subject_types_supported":["public","pairwise"],"id_token_signing_alg_values_supported":["PS384","ES384","RS384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512"],"id_token_encryption_alg_values_supported":["RSA-OAEP","RSA-OAEP-256","RSA1_5"],"id_token_encryption_enc_values_supported":["A256GCM","A192GCM","A128GCM","A128CBC-HS256","A192CBC-HS384","A256CBC-HS512"],"userinfo_signing_alg_values_supported":["PS384","ES384","RS384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512","none"],"userinfo_encryption_alg_values_supported":["RSA-OAEP","RSA-OAEP-256","RSA1_5"],"userinfo_encryption_enc_values_supported":["A256GCM","A192GCM","A128GCM","A128CBC-HS256","A192CBC-HS384","A256CBC-HS512"],"request_object_signing_alg_values_supported":["PS384","ES384","RS384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512","none"],"request_object_encryption_alg_values_supported":["RSA-OAEP","RSA-OAEP-256","RSA1_5"],"request_object_encryption_enc_values_supported":["A256GCM","A192GCM","A128GCM","A128CBC-HS256","A192CBC-HS384","A256CBC-HS512"],"response_modes_supported":["query","fragment","form_post","query.jwt","fragment.jwt","form_post.jwt","jwt"],"registration_endpoint":"http://localhost:8080/realms/supr/clients-registrations/openid-connect","token_endpoint_auth_methods_supported":["private_key_jwt","client_secret_basic","client_secret_post","tls_client_auth","client_secret_jwt"],"token_endpoint_auth_signing_alg_values_supported":["PS384","ES384","RS384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512"],"introspection_endpoint_auth_methods_supported":["private_key_jwt","client_secret_basic","client_secret_post","tls_client_auth","client_secret_jwt"],"introspection_endpoint_auth_signing_alg_values_supported":["PS384","ES384","RS384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512"],"authorization_signing_alg_values_supported":["PS384","ES384","RS384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512"],"authorization_encryption_alg_values_supported":["RSA-OAEP","RSA-OAEP-256","RSA1_5"],"authorization_encryption_enc_values_supported":["A256GCM","A192GCM","A128GCM","A128CBC-HS256","A192CBC-HS384","A256CBC-HS512"],"claims_supported":["aud","sub","iss","auth_time","name","given_name","family_name","preferred_username","email","acr"],"claim_types_supported":["normal"],"claims_parameter_supported":true,"scopes_supported":["openid","email","phone","web-origins","microprofile-jwt","address","profile","roles","offline_access","acr"],"request_parameter_supported":true,"request_uri_parameter_supported":true,"require_request_uri_registration":true,"code_challenge_methods_supported":["plain","S256"],"tls_client_certificate_bound_access_tokens":true,"revocation_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/revoke","revocation_endpoint_auth_methods_supported":["private_key_jwt","client_secret_basic","client_secret_post","tls_client_auth","client_secret_jwt"],"revocation_endpoint_auth_signing_alg_values_supported":["PS384","ES384","RS384","HS256","HS512","ES256","RS256","HS384","ES512","PS256","PS512","RS512"],"backchannel_logout_supported":true,"backchannel_logout_session_supported":true,"device_authorization_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/auth/device","backchannel_token_delivery_modes_supported":["poll","ping"],"backchannel_authentication_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/ext/ciba/auth","backchannel_authentication_request_signing_alg_values_supported":["PS384","ES384","RS384","ES256","RS256","ES512","PS256","PS512","RS512"],"require_pushed_authorization_requests":false,"pushed_authorization_request_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/ext/par/request","mtls_endpoint_aliases":{"token_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/token","revocation_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/revoke","introspection_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/token/introspect","device_authorization_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/auth/device","registration_endpoint":"http://localhost:8080/realms/supr/clients-registrations/openid-connect","userinfo_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/userinfo","pushed_authorization_request_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/ext/par/request","backchannel_authentication_endpoint":"http://localhost:8080/realms/supr/protocol/openid-connect/ext/ciba/auth"}}`

I am also able to generate the JWT access token and passing it while accessing the API but keycloak seems to be not allowing to fetch jwk-set.

Can anyone help what is going wrong here?

**Comment by embesozzi** [comment]
Enabling OAuth2 debugging in your Spring Boot application can provide more detailed information about authorization issues related to OAuth2. You can add the following configuration to your `application.yml` file:

```yaml
logging:
  level:
    org.springframework.security.oauth2: DEBUG
```

OR

```yaml
logging:
  level:
    org.springframework.security: DEBUG
```

**Comment by embesozzi** [comment]
On the other hand, to validate the access token (AT), the API acting as an OAuth2 Resource Server will retrieve the public keys from the `jwks_uri` claim in OpenID Connect Discovery configuration. Therefore, it's important to check if the API is capable of reaching the endpoint `http://localhost:8080/realms/supr/protocol/openid-connect/certs`.
